ffi/blitbuffer.lua

```diff
@@ -1826,6 +1826,99 @@ function BB_mt.__index:writePNGFromBGR(filename)
     C.free(cdata)
 end
 
+--jit.opt.start(1)
+--jit.opt.start("loopunroll=10")
+function BB_mt.__index:writeBMP(filename)
+    local function write_uint32(of, data)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(data))
+    end
+
+    local output_channels = 3
+    local w, h = self:getWidth(), self:getHeight()
+    local stride = w * output_channels
+
+    local of, err = io.open(filename, "wb")
+    if err ~= nil then
+        return err
+    end
+
+    local bbdump
+    local data
+    if self:getType() == TYPE_BBRGB24 then
+        data = ffi.cast("unsigned char *", self.data)
+    else
+        bbdump = BB.new(w, h, TYPE_BBRGB24, nil)
+        bbdump:blitFrom(self)
+        data = ffi.cast("unsigned char *", bbdump.data)
+    end
+
+    local filesize = stride * h + 54
+    -- update filesize, if stride is not a multiple of 4
+    if band(stride, 3) ~= 0 then
+        filesize = filesize + (4 - band(stride,3))*h
```

Comment:
You could also remember the amount of padding bytes computed here (i.e., `4 - (stride & 3)`) and then use that later when you need to fill-in the padding bytes at the end of each row, instead of re-computing it based on `pos`.

(This becomes a moot point if you switch to a `calloc` buffer, since it'll be zero-initialized, so, it'll come pre-padded already ;p).

Reply:
This would be an optimization, but maybe it is already done by CSE.